api/memory.go

```diff
@@ -8,7 +8,22 @@ import "C"
 import "unsafe"
 
 func allocateRust(data []byte) C.Buffer {
-	return C.allocate_rust(u8_ptr(unsafe.Pointer(&data[0])), usize(len(data)))
+	var ret C.Buffer
+	if data == nil {
+		// Just return a null buffer
+		ret = C.Buffer{
+			ptr: u8_ptr(nil),
+			len: usize(0),
+			cap: usize(0),
+		}
+	} else if len(data) == 0 {
+		// This will create an empty vector
```

Comment:
I'd love to see a test case here to see if this differs from the `nil` case. Do we get a real pointer here?

Reply:
Yes, we do. in Rust, the pointer help by a slice or vector may never be null, so it stores the alignment of the generic type instead (a power of 2,  >=1). When requesting an empty vector (e.g. `vec![]` or `Vec::new()`) it doesn't allocate a buffer, and uses that technique. You can see this by analyzing the structure of an empty vector with different types and lengths:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=30cfc49d39e4aebee1ff763c0efd8200

Reply:
later when this `Buffer` instance is returned to Rust, it checks if the `ptr` field is null. if it isn't it can safely `consume` the `Buffer`, even if the vector is empty

Reply:
This is a performance optimization made when creating vectors in Rust as any capacity increase will require using a new memory region anyways. But do we need to do this optimization as well? Why not just drop this code branch in do

```
ret = C.allocate_rust(u8_ptr(unsafe.Pointer(&data[0])), usize(len(data)))
```

in all cases?

Reply:
because running the `data[0]` part causes a Go panic when the buffer is empty:
https://play.golang.org/p/xvDY3g9OqUk

Reply:
We could create the empty Vector like rust would, but we will have no guarantee that it will keep using this same representation in the future. That's why i prefer to keep it this way to prevent breakages

Reply:
Perfect, thanks. Can you add this as a comment to the special treatment of `len(data) == 0`? This is Go internals that are easy to miss in a refactoring. E.g. in C++ (legazy code at least), you always use stuff like `&mystring.data[0]` to refer to the beginning of a potentially empty string.

Reply:
I agree :smile: , I missed it too at first, and one of the commits here is a fix for that.
I'll add the comment tomorrow morning.

Reply:
comment added in c40b267